File: backend/api/feedback.py
from fastapi import APIRouter, Depends, HTTPException, Query, Request, status
from sqlalchemy.ext.asyncio import AsyncSession

from api.deps import get_current_user, require_admin
from core.database import get_db
from core.limiter import limiter
from models.user import User
from schemas.feedback import (
    LikeResponse,
    PublicFeedbackItem,
    PublicFeedbackListResponse,
    UserFeedbackListResponse,
    UserFeedbackRequest,
    UserFeedbackItem,
)
from services.feedback_service import (
    count_public_feedback,
    list_public_feedback,
    list_user_feedback,
    submit_user_feedback,
    toggle_feedback_like,
)
from services.audit_log_service import write_audit_log

router = APIRouter(prefix="/feedback", tags=["feedback"])


@router.post("", status_code=status.HTTP_201_CREATED)
@limiter.limit("10/minute")
async def create_user_feedback(
    request: Request,
    data: UserFeedbackRequest,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """用户提交意见箱反馈。

    限流 10 条/分钟，防止刷反馈。
    """
    fb = await submit_user_feedback(
        db,
        user_id=current_user.id,
        content=data.content,
        feedback_type=data.type,
    )
    await write_audit_log(
        db,
        user_id=current_user.id,
        action="feedback_submit",
        target_type="feedback",
        target_id=str(fb.id),
        detail={"result": "success", "request_id": request.headers.get("X-Request-ID"), "type": data.type},
        ip=request.client.host if request.client else None,
    )
    return {"id": fb.id, "detail": "反馈已提交，感谢你的建议"}


@router.get("", response_model=UserFeedbackListResponse)
async def get_user_feedback_list(
    limit: int = Query(20, ge=1, le=100),
    offset: int = Query(0, ge=0),
    db: AsyncSession = Depends(get_db),
    _admin: User = Depends(require_admin),
):
    """管理员分页查看意见箱反馈。

    仅 settings.ADMIN_EMAILS 中的用户可访问。
    """
    items, total = await list_user_feedback(db, limit=limit, offset=offset)
    return UserFeedbackListResponse(
        items=[UserFeedbackItem.model_validate(it) for it in items],
        total=total,
    )


# ── 公开接口（所有登录用户可访问） ──────────────────────


@router.get("/public", response_model=PublicFeedbackListResponse)
async def get_public_feedback_list(
    limit: int = Query(50, ge=1, le=100),
    offset: int = Query(0, ge=0),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """公开反馈列表 — 所有登录用户可看。返回用户名 + 点赞数 + 是否已点赞。"""
    items = await list_public_feedback(db, current_user.id, limit=limit, offset=offset)
    total = await count_public_feedback(db)
    return PublicFeedbackListResponse(
        items=[PublicFeedbackItem(**item) for item in items],
        total=total,
    )


@router.post("/public/{fb_id}/like", response_model=LikeResponse)
async def like_feedback(
    request: Request,
    fb_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """点赞/取消点赞反馈（toggle）。"""
    try:
        likes_count, is_liked = await toggle_feedback_like(db, current_user.id, fb_id)
    except LookupError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="反馈不存在",
        )
    await write_audit_log(
        db,
        user_id=current_user.id,
        action="feedback_like_toggle",
        target_type="feedback",
        target_id=str(fb_id),
        detail={
            "result": "success",
            "request_id": request.headers.get("X-Request-ID"),
            "is_liked": is_liked,
        },
        ip=request.client.host if request.client else None,
    )
    return LikeResponse(likes_count=likes_count, is_liked=is_liked)

File: frontend/src/api/feedback.ts
import { api } from "./client";

// ── 类型定义 ──

export interface UserFeedbackItem {
  id: number;
  user_id: number;
  content: string;
  type: string;
  status: string;
  created_at: string;
}

export interface PublicFeedbackItem {
  id: number;
  content: string;
  type: string;
  status: string;
  created_at: string;
  user_display: string;
  likes_count: number;
  is_liked: boolean;
}

export interface PublicFeedbackListResponse {
  items: PublicFeedbackItem[];
  total: number;
}

// ── 反馈类型 ──

export const FEEDBACK_TYPES = [
  { value: "bug", label: "Bug 反馈" },
  { value: "feature", label: "功能建议" },
  { value: "ux", label: "体验优化" },
  { value: "other", label: "其他" },
] as const;

// ── API 函数 ──

/** 提交反馈 */
export async function submitFeedback(
  content: string,
  type: string
): Promise<{ id: number; detail: string }> {
  return api.post("/api/v1/feedback", { content, type }) as Promise<{
    id: number;
    detail: string;
  }>;
}

/** 公开反馈列表（所有登录用户可看） */
export async function listPublicFeedback(
  limit = 50,
  offset = 0
): Promise<PublicFeedbackListResponse> {
  return api.get(
    `/api/v1/feedback/public?limit=${limit}&offset=${offset}`
  ) as Promise<PublicFeedbackListResponse>;
}

/** 点赞/取消点赞反馈（toggle） */
export async function toggleFeedbackLike(
  feedbackId: number
): Promise<{ likes_count: number; is_liked: boolean }> {
  return api.post(
    `/api/v1/feedback/public/${feedbackId}/like`
  ) as Promise<{ likes_count: number; is_liked: boolean }>;
}

/** 管理员查看反馈列表 */
export async function listFeedback(
  limit = 20,
  offset = 0
): Promise<{ items: UserFeedbackItem[]; total: number }> {
  return api.get(
    `/api/v1/feedback?limit=${limit}&offset=${offset}`
  ) as Promise<{ items: UserFeedbackItem[]; total: number }>;
}
